### ISSUE TITLE
Kernel panic with OSXFUSE 3.5.2 (Spotlight)

### ISSUE THREAD

**Issue by alonre** [osxfuse/osxfuse#330]
Forked from #232 (but different).

FUSE 3.5.2 on macOS 10.12.
Clue: the panic occurred when I woke the machine up after it has been asleep for a short while.
Indexing was enabled on the volume at the time (was turned on ~1h earlier).
We saw the same thing about 18 months ago on a previous attempt to support Spotlight indexing on our Virtual FS.

```
Tue Oct 25 12:15:09 2016

*** Panic Report ***
panic(cpu 4 caller 0xffffff7f9d2aca2f): "osxfuse: Attempted to lock max-locked recursive lock."@/Users/benjamin/Documents/Projekte/FUSE/Repositories/osxfuse/kext/osxfuse/fuse_locking.c:432
Backtrace (CPU 4), Frame : Return Address
0xffffff922ae3b380 : 0xffffff8019cf748c 
0xffffff922ae3b400 : 0xffffff7f9d2aca2f 
0xffffff922ae3b420 : 0xffffff7f9d2b28ba 
0xffffff922ae3b440 : 0xffffff801a201d0d 
0xffffff922ae3b530 : 0xffffff8019d61108 
0xffffff922ae3b580 : 0xffffff8019d60e93 
0xffffff922ae3b5c0 : 0xffffff8019d6fd13 
0xffffff922ae3b6f0 : 0xffffff8019d749e4 
0xffffff922ae3b8d0 : 0xffffff8019e098da 
0xffffff922ae3bab0 : 0xffffff8019ca9743 
0xffffff922ae3bad0 : 0xffffff8019ca684b 
0xffffff922ae3bc30 : 0xffffff801a18dc44 
0xffffff922ae3bc70 : 0xffffff7f9d2ba048 
0xffffff922ae3bd30 : 0xffffff7f9d2b3134 
0xffffff922ae3bd60 : 0xffffff8019f53fe0 
0xffffff922ae3bde0 : 0xffffff8019f4570e 
0xffffff922ae3be50 : 0xffffff801a1a3927 
0xffffff922ae3bef0 : 0xffffff801a1a3ba2 
0xffffff922ae3bf50 : 0xffffff801a22a366 
0xffffff922ae3bfb0 : 0xffffff8019ca9f46 
      Kernel Extensions in backtrace:
         com.github.osxfuse.filesystems.osxfuse(3.5.2)[FAD7BE87-C14C-345F-A049-8689D1F080A0]@0xffffff7f9d2ab000->0xffffff7f9d2c3fff

BSD process name corresponding to current thread: mds_stores

Mac OS version:
16A323

Kernel version:
Darwin Kernel Version 16.0.0: Mon Aug 29 17:56:20 PDT 2016; root:xnu-3789.1.32~3/RELEASE_X86_64
Kernel UUID: 622D2470-C34D-31F9-A62B-6AA9A3C6A3CD
Kernel slide:     0x0000000019a00000
Kernel text base: 0xffffff8019c00000
__HIB  text base: 0xffffff8019b00000
System model name: MacBookPro11,3 (Mac-2BD1B31983FE1663)

System uptime in nanoseconds: 190506610788775
last loaded kext at 183690286028621: com.getdropbox.dropbox.kext    1.8.1 (addr 0xffffff7f9d2c4000, size 49152)
last unloaded kext at 183689454059256: com.getdropbox.dropbox.kext  1.8.1 (addr 0xffffff7f9d2c4000, size 24576)
loaded kexts:
com.getdropbox.dropbox.kext 1.8.1
com.github.osxfuse.filesystems.osxfuse  3.5.2
com.vmware.kext.vmioplug.15.1.7 15.1.7
com.vmware.kext.vmx86   0377.10.13
com.vmware.kext.vmnet   0377.10.13
com.vmware.kext.vsockets    90.8.0
com.vmware.kext.vmci    90.8.1
org.virtualbox.kext.VBoxNetAdp  5.0.26
org.virtualbox.kext.VBoxNetFlt  5.0.26
org.virtualbox.kext.VBoxUSB 5.0.26
org.virtualbox.kext.VBoxDrv 5.0.26
com.apple.filesystems.smbfs 3.1
com.apple.driver.AppleHWSensor  1.9.5d0
com.apple.driver.AudioAUUC  1.70
com.apple.filesystems.autofs    3.0
com.apple.driver.AGPM   110.23.7
com.apple.driver.ApplePlatformEnabler   2.7.0d0
com.apple.driver.X86PlatformShim    1.0.0
com.apple.driver.AppleOSXWatchdog   1
com.apple.driver.AppleGraphicsDevicePolicy  3.13.60
com.apple.driver.AppleHDA   276.26
com.apple.driver.AppleUpstreamUserClient    3.6.4
com.apple.GeForce   10.1.2
com.apple.driver.AppleSMCLMU    208
com.apple.driver.AppleIntelHD5000Graphics   10.1.8
com.apple.driver.AppleMCCSControl   1.2.13
com.apple.driver.AppleCameraInterface   5.57.0
com.apple.driver.AppleMuxControl    3.13.60
com.apple.driver.AppleIntelFramebufferAzul  10.1.8
com.apple.iokit.BroadcomBluetoothHostControllerUSBTransport 5.0.0f18
com.apple.driver.pmtelemetry    1
com.apple.iokit.IOUserEthernet  1.0.1
com.apple.iokit.IOBluetoothSerialManager    5.0.0f18
com.apple.Dont_Steal_Mac_OS_X   7.0.0
com.apple.driver.AppleHV    1
com.apple.driver.AppleIntelSlowAdaptiveClocking 4.0.0
com.apple.driver.AppleLPC   3.1
com.apple.driver.AppleThunderboltIP 3.0.8
com.apple.driver.AppleUSBCardReader 404.1.1
com.apple.driver.AppleUSBTCButtons  251
com.apple.driver.AppleUSBTCKeyboard 251
com.apple.AppleFSCompression.AppleFSCompressionTypeDataless 1.0.0d1
com.apple.AppleFSCompression.AppleFSCompressionTypeZlib 1.0.0
com.apple.BootCache 39
com.apple.filesystems.hfs.kext  366.1.1
com.apple.iokit.IOAHCIBlockStorage  295.1.1
com.apple.driver.AppleAHCIPort  326
com.apple.driver.AirPort.Brcm4360   1100.37.1a16
com.apple.driver.AppleSmartBatteryManager   161.0.0
com.apple.driver.AppleACPIButtons   5.0
com.apple.driver.AppleRTC   2.0
com.apple.driver.AppleHPET  1.8
com.apple.driver.AppleSMBIOS    2.1
com.apple.driver.AppleACPIEC    5.0
com.apple.driver.AppleAPIC  1.7
com.apple.nke.applicationfirewall   171
com.apple.security.quarantine   3
com.apple.security.TMSafetyNet  8
com.apple.driver.AppleBluetoothHIDKeyboard  197
com.apple.driver.AppleHIDKeyboard   197
com.apple.driver.IOBluetoothHIDDriver   5.0.0f18
com.apple.iokit.IOUSBUserClient 900.4.1
com.apple.kext.triggers 1.0
com.apple.driver.DspFuncLib 276.26
com.apple.kext.OSvKernDSPLib    525
com.apple.nvidia.driver.NVDAGK100Hal    10.1.2
com.apple.nvidia.driver.NVDAResman  10.1.2
com.apple.driver.AppleSMBusController   1.0.14d1
com.apple.driver.AppleHDAController 276.26
com.apple.iokit.IOHDAFamily 276.26
com.apple.iokit.IOAudioFamily   205.11
com.apple.vecLib.kext   1.2.0
com.apple.driver.AppleBacklightExpert   1.1.0
com.apple.iokit.IONDRVSupport   2.4.1
com.apple.driver.AppleGraphicsControl   3.13.60
com.apple.AppleGraphicsDeviceControl    3.13.60
com.apple.iokit.IOAcceleratorFamily2    288.13
com.apple.driver.X86PlatformPlugin  1.0.0
com.apple.iokit.IOBluetoothHostControllerUSBTransport   5.0.0f18
com.apple.iokit.IOBluetoothHostControllerTransport  5.0.0f18
com.apple.iokit.IOSurface   152
com.apple.iokit.IOGraphicsFamily    2.4.1
com.apple.iokit.IOSerialFamily  11
com.apple.iokit.IOBluetoothFamily   5.0.0f18
com.apple.driver.AppleSMC   3.1.9
com.apple.iokit.IOSlowAdaptiveClockingFamily    1.0.0
com.apple.driver.IOPlatformPluginFamily 6.0.0d8
com.apple.iokit.IOSCSIBlockCommandsDevice   394
com.apple.iokit.IOUSBMassStorageDriver  131.1.1
com.apple.iokit.IOSCSIArchitectureModelFamily   394
com.apple.driver.AppleUSBMultitouch 258
com.apple.driver.usb.IOUSBHostHIDDevice 1.1
com.apple.driver.usb.networking 5.0.0
com.apple.driver.usb.AppleUSBHostCompositeDevice    1.1
com.apple.driver.usb.AppleUSBHub    1.1
com.apple.driver.CoreStorage    540
com.apple.filesystems.hfs.encodings.kext    1
com.apple.iokit.IOAHCIFamily    288
com.apple.driver.AppleThunderboltDPInAdapter    4.5.3
com.apple.driver.AppleThunderboltDPAdapterFamily    4.5.3
com.apple.driver.AppleThunderboltPCIDownAdapter 2.0.3
com.apple.driver.AppleThunderboltNHI    4.1.3
com.apple.iokit.IOThunderboltFamily 6.2.1
com.apple.iokit.IO80211Family   1200.12.2
com.apple.driver.mDNSOffloadUserClient  1.0.1b8
com.apple.iokit.IONetworkingFamily  3.2
com.apple.driver.corecapture    1.0.4
com.apple.driver.usb.AppleUSBXHCIPCI    1.1
com.apple.driver.usb.AppleUSBXHCI   1.1
com.apple.driver.usb.AppleUSBHostPacketFilter   1.0
com.apple.iokit.IOUSBFamily 900.4.1
com.apple.iokit.IOUSBHostFamily 1.1
com.apple.driver.AppleUSBHostMergeProperties    1.1
com.apple.driver.AppleEFINVRAM  2.1
com.apple.driver.AppleBusPowerController    1.0
com.apple.iokit.IOHIDFamily 2.0.0
com.apple.driver.AppleEFIRuntime    2.0
com.apple.iokit.IOSMBusFamily   1.1
com.apple.security.sandbox  300.0
com.apple.kext.AppleMatch   1.0.0d1
com.apple.driver.AppleKeyStore  2
com.apple.driver.AppleMobileFileIntegrity   1.0.5
com.apple.driver.AppleCredentialManager 1.0
com.apple.driver.DiskImages 444
com.apple.iokit.IOStorageFamily 2.1
com.apple.iokit.IOReportFamily  31
com.apple.driver.AppleFDEKeyStore   28.30
com.apple.driver.AppleACPIPlatform  5.0
com.apple.iokit.IOPCIFamily 2.9
com.apple.iokit.IOACPIFamily    1.4
com.apple.kec.Libm  1
com.apple.kec.pthread   1
com.apple.kec.corecrypto    1.0
```

**Comment by alonre** [comment]
Copied comment by @bfleischer:

> It looks like this is a different issue from the one @tv42 reported. But it should be easy to fix.

> Symbolicated first stack trace:
```
panic(cpu 4 caller 0xffffff7f9d2aca2f): "osxfuse: Attempted to lock max-locked recursive lock."@/Users/benjamin/Documents/Projekte/FUSE/Repositories/osxfuse/kext/osxfuse/fuse_locking.c:432
Backtrace (CPU 4), Frame : Return Address
0xffffff922ae3b380 : 0xffffff8019cf748c 
0xffffff922ae3b400 : 0xffffff7f9d2aca2f fusefs_recursive_lock_lock (in osxfuse) (fuse_locking.c:434)
0xffffff922ae3b420 : 0xffffff7f9d2b28ba fuse_biglock_vnop_pagein (in osxfuse) (fuse_biglock_vnops.c:420)
0xffffff922ae3b440 : 0xffffff801a201d0d vnode_pagein (in kernel) (kpi_vfs.c:5165)
0xffffff922ae3b530 : 0xffffff8019d61108 vnode_pager_cluster_read (in kernel) (bsd_vm.c:877)
0xffffff922ae3b580 : 0xffffff8019d60e93 vnode_pager_data_request (in kernel) (bsd_vm.c:645)
0xffffff922ae3b5c0 : 0xffffff8019d6fd13 vm_fault_page (in kernel) (memory_object.c:2228)
0xffffff922ae3b6f0 : 0xffffff8019d749e4 vm_fault_internal (in kernel) (vm_fault.c:4686)
0xffffff922ae3b8d0 : 0xffffff8019e098da kernel_trap (in kernel) (vm_fault.c:3397)
0xffffff922ae3bab0 : 0xffffff8019ca9743 trap_from_kernel (in kernel) + 38
0xffffff922ae3bad0 : 0xffffff8019ca684b _bcopy (in kernel) + 11
0xffffff922ae3bc30 : 0xffffff801a18dc44 uiomove64 (in kernel) (kern_subr.c:181)
0xffffff922ae3bc70 : 0xffffff7f9d2ba048 fuse_vnop_write (in osxfuse) (fuse_vnops.c:3899)
0xffffff922ae3bd30 : 0xffffff7f9d2b3134 fuse_biglock_vnop_write (in osxfuse) (fuse_biglock_vnops.c:727)
0xffffff922ae3bd60 : 0xffffff8019f53fe0 VNOP_WRITE (in kernel) (kpi_vfs.c:3408)
0xffffff922ae3bde0 : 0xffffff8019f4570e vn_write (in kernel) (vfs_vnops.c:1150)
0xffffff922ae3be50 : 0xffffff801a1a3927 dofilewrite (in kernel) (kern_descrip.c:5781)
0xffffff922ae3bef0 : 0xffffff801a1a3ba2 pwrite_nocancel (in kernel) (sys_generic.c:551)
0xffffff922ae3bf50 : 0xffffff801a22a366 unix_syscall64 (in kernel) (systemcalls.c:376)
0xffffff922ae3bfb0 : 0xffffff8019ca9f46 hndl_unix_scall64 (in kernel) + 22
      Kernel Extensions in backtrace:
         com.github.osxfuse.filesystems.osxfuse(3.5.2)[FAD7BE87-C14C-345F-A049-8689D1F080A0]@0xffffff7f9d2ab000->0xffffff7f9d2c3fff
```

**Comment by bfleischer** [comment]
The issue should be fixed in commit https://github.com/osxfuse/kext/commit/899ef41f25b00c70677520f24ffdfecef451869e. The fix will be in version 3.5.4. Thanks again for reporting the issue.
